------------------------------------------------------------------------

title: "Myd88 Hershey Collaboration on 16S Analysis - Processing with phyloseq" output: html_document date: "2024-01-10" author: "Hanh Tran" #R version: 4.2.2 (2022-10-31)

------------------------------------------------------------------------

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
getwd()
```

#### First we load all the neccessary packages and define path variables

```{r}
library("knitr")
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("phyloseq")
library('phyloseq')
library('dada2')
library("dendextend") # load the package
library("DESeq2")
library("ggplot2")
library("MicrobiotaProcess")
library("ggtree")
library("vegan")
library("phangorn")
library("DECIPHER")
library("plyr"); packageVersion("plyr")
library("tidyverse"); packageVersion("tidyverse")
library("phyloseq"); packageVersion("phyloseq")
library("vegan"); packageVersion("vegan")
library("gridExtra"); packageVersion("gridExtra")
library("knitr"); packageVersion("knitr")
library("DESeq2"); packageVersion("DESeq2")
library("plotly"); packageVersion("plotly")
library("microbiome"); packageVersion("microbiome")
library("ggpubr"); packageVersion("ggpubr")
library("data.table"); packageVersion("data.table")
library("ALDEx2")
library("RColorBrewer")
library("ANCOMBC")
library('gghalves')
library('ggalluvial')
library('ggtreeExtra')
library('microbiomeMarker')


```

## Set global ggplot2 theme and options

### This sets the plotting aesthetics for every ggplot2 for the rest of the document. There are a tremendous number of ways to customize your ggplot2 settings using theme_set (see: <http://ggplot2.tidyverse.org/reference/theme_get.html>). It is best practice to do this at the beginning of the RMarkdown document so that these settings propagated to every plot through the rest of the document.

```{r global-theme-settings, include=FALSE}
# Set global theming
# This theme set will change the ggplot2 defaults to use the b&w settings (removes the default gray background) and sets the default font to 12pt Arial
theme_set(theme_bw(base_size = 12))

# The palette with grey:
cbPalette <- c("#999999","#E69F00","#56B4E9","#009E73",
               "#F0E442","#0072B2","#D55E00","#CC79A7",
               "#003366", "#660066","#006666", "indianred4")
```

## Read in your data

### The output standard dada2 workflow will be an RDS file. In this case the file is called *ASV_physeq*. We will set the phyloseq object as ps0. (ps is shorthand for PhyloSeq. The 0 indicates it is the 'base' version of the file. As it is modified this can be changed to ps1, ps2, etc.).

```{r initiate-data}
# Read in an RDS file containing taxonomic and count information
ps0 <- readRDS("./RDS/ASV_physeq.RDS")
ps0

# Perform a few sanity checks
sample_variables(ps0) # Display variables from the mapping file
ntaxa(ps0) # Total number of taxa in the entire data
rank_names(ps0) # Taxonomic ranks
get_taxa_unique(ps0, "Phylum") # Unique phylum names in the file

```

### Data summary and assessment

#### While there are numerous possible ways to evaluate your data, a standard starting approach would consist of the following steps:

#### 1) Evaluate Amplicon Sequence Variants (ASV) summary statistics

#### 2) Detect and remove outlier samples

#### 3) Taxon cleaning

#### 4) Prevalence estimation and filtering

#### *Step 1: Evaluate Amplicon Sequence Variants (ASV) summary statistics*

#### Begin by running the following R chunk to produce ASV summary plots.

```{r data-assessment}
# Create a new data frame of the sorted row sums, a column of sorted values from 1 to the total number of individuals/counts for each ASV and a categorical variable stating these are all ASVs.
readsumsdf <- data.frame(nreads = sort(taxa_sums(ps0), decreasing = TRUE), 
                        sorted = 1:ntaxa(ps0),
                        type = "ASVs")

# Make a data frame with a column for the read counts of each sample for histogram production
sample_sum_df <- data.frame(sum = sample_sums(ps0))

# Make plots
# Generates a bar plot with # of reads (y-axis) for each taxa. Sorted from most to least abundant
# Generates a second bar plot with # of reads (y-axis) per sample. Sorted from most to least
p.reads = ggplot(readsumsdf, aes(x = sorted, y = nreads)) +
  geom_bar(stat = "identity") +
  ggtitle("ASV Assessment") +
  scale_y_log10() +
  facet_wrap(~type, scales = "free") +
  ylab("# of Sequences")

# Histogram of the number of Samples (y-axis) at various read depths
p.reads.hist <- ggplot(sample_sum_df, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "firebrick3", binwidth = 150) +
  ggtitle("Distribution of sample sequencing depth") + 
  xlab("Read counts") +
  ylab("# of Samples")

# Final plot, side-by-side
grid.arrange(p.reads, p.reads.hist, ncol = 2)

# Basic summary statistics
summary(sample_sums(ps0))
```

### The above data assessment is useful for getting an idea of 1) the number of sequences per taxa (left plot). This will normally be a "long tail" with some taxa being highly abundant in the data tapering off to taxa with very few reads, 2) the number of reads per sample. Note the spike at the lowest number of reads due to samples taken from mice given antibiotics. Very low read count can be indicative of a failed reaction. Both of these plots will help give an understanding of how your data are structured across taxa and samples and will vary depending on the nature of your samples.

### Samples with unexpectedly low number of sequences can be considered for removal. This is an intuitive process and should be instructed by your understanding of the samples in your study. For example, if you have 5 samples from stool samples, one would expect to obtain thousands, if not several thousands of ASV. This may not be the case for other tissues, such as spinal fluid or tissue samples. Similarly, you may not expect thousands of ASV from samples obtained from antibiotic treated organisms. Following antibiotic treatment you may be left with only dozens or hundreds of ASV. So contextual awareness about the biology of your system should guide your decision to remove samples based on ASV number.

### Importantly, at each stage you should document and justify your decisions. If you are concerned that sample removal will alter the interpretation of your results, you should run your analysis on the full data and the data with the sample(s) removed to see how the decision affects your interpretation.

### The above plots provide overall summaries about the number of ASV found in all of your samples. However, they are not very useful for identifying and removing specific samples. One way to do this is using code from the following R chunk.

## *Step 2: Detect and remove outlier samples*

## Detecting and potentially removing samples outliers (those samples with underlying data that do not conform to experimental or biological expectations) can be useful for minimizing technical variance. One way to identify sample outliers is shown in the R chunk below.

```{r sample-removal-identification}
# Format a data table to combine sample summary data with sample variable data
ss <- sample_sums(ps0)
sd <- as.data.frame(sample_data(ps0))
ss.df <- merge(sd, data.frame("ASV" = ss), by ="row.names")


# Plot the data by the treatment variable
y = 1000 # Set a threshold for the minimum number of acceptable reads. Can start as a guess
x = "Isolation.Source" # Set the x-axis variable you want to examine
label = "Sample_name" # This is the label you want to overlay on the points

p.ss.boxplot <- ggplot(ss.df, aes_string(x, y = "ASV", color = "Treatment")) + 
  geom_boxplot(outlier.colour="NA", position = position_dodge(width = 0.8)) +
  geom_jitter(size = 2, alpha = 0.6) +
  scale_y_log10() +
  facet_wrap(~Treatment) +
  geom_hline(yintercept = y, lty = 2) +
  geom_text(aes_string(label = label), size = 3, nudge_y = 0.05, nudge_x = 0.05)
p.ss.boxplot

```

### The data does not have samples with fewer than 1,000 ASV in the treatment groups. So we don't need to remove any outlier sample

### *Step 3: Taxon cleaning*

#### The following R chunk removes taxa not-typically part of a bacterial microbiome analysis.

```{r taxon-cleaning}
# Some examples of taxa you may not want to include in your analysis
get_taxa_unique(ps0, "Kingdom")
get_taxa_unique(ps0, "Class")

ps0 # Check the number of taxa prior to removal
ps1 <- ps0 %>%
  subset_taxa(
    Kingdom == "Bacteria" &
    Family  != "mitochondria" &
    Class   != "Chloroplast" &
    Phylum != "Cyanobacteria/Chloroplast"
  )
ps1 # Confirm that the taxa were removed
get_taxa_unique(ps1, "Kingdom")
get_taxa_unique(ps1, "Class")

rank_names(ps0)
# Create table, number of features for each phyla
table(phyloseq::tax_table(ps1)[, "Phylum"], exclude = NULL)
```

#### Prevalance assessment

##### Identification of taxa that are poorly represented in an unsupervised manner can identify taxa that will have little to no effect on downstream analysis. Sufficient removal of these "low prevalence" features can enhance many analysis by focusing statistical testing on taxa common throughout the data. This approach is frequently poorly documented or justified in methods sections of manuscripts, but will typically read something like, "Taxa present in fewer than 3% of all of the samples within the study population and less than 5% relative abundance were removed from all subsequent analysis. While the ultimate selection criteria can still be subjective, the following plots can be useful for making your selection criteria.

```{r prevalence-assessment}
# Prevalence estimation
# Calculate feature prevalence across the data set
prevdf <- apply(X = otu_table(ps1),MARGIN = ifelse(taxa_are_rows(ps1), yes = 1, no = 2),FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to prevdf
prevdf <- data.frame(Prevalence = prevdf, TotalAbundance = taxa_sums(ps1), phyloseq::tax_table(ps1))

plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

#Prevalence plot
prevdf1 <- subset(prevdf, Phylum %in% get_taxa_unique(ps0, "Phylum"))
p.prevdf1 <- ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(ps1),color=Family)) +
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +
  geom_point(size = 3, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) +
  theme(legend.position="none") +
  ggtitle("Phylum Prevalence in All Samples\nColored by Family")
p.prevdf1

```

#### A useful next step is to explore feature prevalence in the dataset, which we will define here as the number of samples in which a taxon appears at least once. This code will produce a plot of all of the Phyla present in your samples along with information about their prevalence (fraction of samples they are present in) and total abundance across all samples. In this example we drew a dashed horizontal line to cross at the 5% prevalence level (present in \> 5% of all of the samples in the study). If you set a threshold to remove taxa below that level you can visually see how many and what types of taxa will be removed. Whatever threshold you choose to use it should be well documented within your materials and methods.

#### An example on how to filter low prevalent taxa is below.

```{r prevelance-filtering}


# Remove specific taxa
# Define a list with taxa to remove
filterPhyla = c( "Acidobacteria" )
filterPhyla

get_taxa_unique(ps1, "Phylum")
ps1.prev <- subset_taxa(ps1, !Phylum %in% filterPhyla) 
get_taxa_unique(ps1.prev, "Phylum")

# Removing taxa that fall below 0.5% prevelance
# Define the prevalence threshold
prevalenceThreshold <- 0.005 * nsamples(ps1)
prevalenceThreshold

# Define which taxa fall within the prevalence threshold
#keepTaxa <- rownames(prevdf1)[(prevdf1$Prevalence >= prevalenceThreshold)]

keepTaxa <- rownames(prevdf1)[prevdf1$Prevalence >= floor(prevalenceThreshold)]


ntaxa(ps1)
# Remove those taxa
ps1.prev <- prune_taxa(keepTaxa, ps1)
ntaxa(ps1.prev)

```

### Data transformation

#### Many analysis in community ecology and hypothesis testing benefit from data transformation. Many microbiome data sets do not fit to a normal distribution, but transforming them towards normality may enable more appropriate data for specific statistical tests. The choice of transformation is not straight forward. There is literature on how frequently used transformations affect certain analysis, but every data set may require different considerations. Therefore, it is recommended that you examine the effects of several transformations on your data and explore how they alter your results and interpretation. The R chunk below implements several commonly used transformations in microbiome research and plots their results. Similar to outlier removal and prevalence filtering, your choice should be justified, tested and documented.

```{r data-transform}
# Rarefy data
ps1.rarefied <- rarefy_even_depth(ps1, rngseed=1, replace=F) #22OTUs were removed because they are no longer present in any sample after random subsampling
seq_tab_rarefied <- otu_table(ps1.rarefied, taxa_are_rows = T)
taxatab_rarefied <- phyloseq::tax_table(ps1.rarefied)
sampleda <- sample_data(ps1.rarefied)
  
# Transform to Relative abundances
ps1.ra <- transform_sample_counts(ps1, function(OTU) OTU/sum(OTU))

# Transform to Proportional Abundance
ps1.prop <- transform_sample_counts(ps1, function(x) min(sample_sums(ps1)) * x/sum(x))

# Log transformation moves to a more normal distribution
ps1.log <- transform_sample_counts(ps1, function(x) log(1 + x))

# View how each function altered count data
par(mfrow=c(1,4))
plot(sort(sample_sums(ps1), TRUE), type = "o", main = "Native", ylab = "RSVs", xlab = "Samples")
plot(sort(sample_sums(ps1.log), TRUE), type = "o", main = "log Transformed", ylab = "RSVs", xlab = "Samples")
plot(sort(sample_sums(ps1.ra), TRUE), type = "o", main = "Relative Abundance", ylab = "RSVs", xlab = "Samples")
plot(sort(sample_sums(ps1.prop), TRUE), type = "o", main = "Proportional Abundance", ylab = "RSVs", xlab = "Samples")
par(mfrow=c(1,4))

# Histograms of the non-transformed data vs. the transformed data can address the shift to normality
p.nolog <- qplot(rowSums(otu_table(ps1))) + ggtitle("Raw Counts") +
  theme_bw() +
  xlab("Row Sum") +
  ylab("# of Samples")

p.log <- qplot(log10(rowSums(otu_table(ps1)))) +
  ggtitle("log10 transformed counts") +
  theme_bw() +
  xlab("Row Sum") +
  ylab("# of Samples")

ggarrange(p.nolog, p.log, ncol = 2, labels = c("A)", "B)"))

```

### Subsetting

#### You will frequently find that you want to analyze a subset of your total data set. There are typically commands that will allow you to do this for each individual analysis, but similar to variable reordering it can sometimes be more convenient to do this towards the beginning of your analysis. This should be done after removal of outlier samples and taxa. If you wish to create transformed versions of each subset you can either subset the transformed data you just generated, or alternatively retransform your subsetted data. The R chunk below is an example subsetting of the example data by treatment.

#### Subsetting away samples can create a situation where taxa are present as empty rows. This is because not every sample has every taxa. These can be removed as shown in the R chunk below.

#### Creating individual subsets like this can be particularly useful when assessing differential abundance using Aldex2.

```{r subsetting}
#Subsets
# Rarefied data:
ps1.cecal.rarefied <- subset_samples(ps1.rarefied, Isolation.Source=="ceacal_content")

ps_cecal_content <- subset_samples(ps1, Isolation.Source=="ceacal_content")

ntaxa(ps1.ra)
ps1.ra <- prune_taxa(taxa_sums(ps1.ra) > 0, ps1.ra)
ntaxa(ps1.ra)

```



### Alpha diversity plotting

#### Alpha diversity is a standard tool researchers can use to calculate the number of bacterial taxa present in a study or study group and the relationships between relative abundance and how evenly taxa are distributed. These are classic representations of species number and diversity in a study which provide useful summary information about the numbers and relative abundances of bacterial taxa within your study.

#### Similar to the plot above, we can calculate several measures of alpha diversity, add them to a data frame and use ggplot2 to follow the alpha diversity trajectory over time.

#### 3.1. Calculation and different analysis of alpha index

###### Alpha index can evaluate the richness and abundance of microbial communities. MicrobiotaProcess provides get_alphaindex to calculate alpha index. Six common diversity measures (Observe, Chao1, ACE, Shannon, Simpson, J) are supported. And the different groups of samples can be tested and visualize by ggbox.

##### Rarefied data ##### 
```{r}
#Convert phyloseq object to mpse
ps1.cecal.rarefied.mpse <- ps1.cecal.rarefied %>% as.MPSE() # cecal content
```

```{r}
# Calculate alpha diversity
alphaobj.cecal_mpse1 <- ps1.cecal.rarefied %>% as.MPSE() # cecal content


alphaobj.cecal_mpse1 %<>% 
    mp_cal_alpha(.abundance=Abundance)


```



```{r}
pdf("./Plots/p_alphaobj.cecal.plots.pdf")
p_alphaobj.cecal <-  mp_plot_alpha(.data = alphaobj.cecal_mpse1,
                                   .group= Treatment,
                                   .alpha = c(ACE, Chao1, Shannon), test = 'wilcox.test') + 
                                   scale_fill_manual(values = c("steelblue","firebrick")) + 
                                   scale_color_manual(values=c("steelblue", "firebrick"), guide="none") +
                                   theme(strip.background = element_rect(colour=NA, fill="grey"),  axis.text.x = element_blank()) 
                                  #labs(title = "Alpha Index of Samples from Colon Content")
p_alphaobj.cecal
dev.off()
```

```{r}
p_alphaobj.cecal
```


### Ordination - Beta Diversity

#### Beta diversity enables you to view overall relationships between samples. These relationships are calculated using a distance metric calculation (of which there are many) and these multi-dimensional relationships are evaluated and viewed in the two dimensions which explain the majority of their variance. Additional dimensions can be explored to evaluate how samples are related to one another.

#### The UniFrac distance metric takes advantage of the phylogenetic relationships between bacterial taxa by down-weighting relationships between bacterial taxa that are phylogenetically related versus ones that are distantly related. Weighted UniFrac builds on this by also taking into account the relative abundances of each taxa. For a more detailed discussion on these distance calculations see: <https://en.wikipedia.org/wiki/UniFrac>. Ultimately, the decision on what distance metric to use should be a result of experimentation and a good understanding of the underlying properties of your data.

#### The following R chunks calculate UniFrac and wUniFrac on a PhyloSeq object and display the the two components of these results that explain a majority of the variance in the data using Principle Coordinates Analysis (PCoA). For a detailed explanation of how PCoA works see: <https://sites.google.com/site/mb3gustame/dissimilarity-based-methods/principal-coordinates-analysis>.

```{r ordination}
#Ordination Analysis
cecal.ord.pcoa.wuni <- ordinate(ps1.cecal.rarefied, method = "PCoA", distance = "wunifrac")

```

#### And now to plot each ordination.


```{r}
## Ordination plots colon only samples
ord.pcoa.wuni.cecal <-  ordinate(ps1.cecal.rarefied, method = "PCoA", distance = "wunifrac")


# Weighted Unifrac
pdf("./Plots/PCoA.weighteduni_cecal.plots.pdf")
p.pcoa.wuni_cecal <- plot_ordination(ps1.cecal.rarefied, ord.pcoa.wuni.cecal, color = "Treatment", axes = c(1,2)) +
  geom_point(size = 2) + scale_color_manual(values = c("steelblue", "firebrick")) +
  geom_text(aes(label = Sample_name), vjust = 1, size = 3) +  # Add point labels
  labs(title = "PCoA of wUniFrac Distances", color = "Treatment") 
p.pcoa.wuni_cecal
dev.off()
```



### Community composition plotting / The visualization of taxonomy abundance

#### Classic bar plots of bacterial phyla present per sample can be useful for communicating "high level" results. These are relatively easy to interpret when major shifts in microbial communities are present, such as in this study where knockouts are used. However, they are not effective at detecting subtle shifts in communities or taxa and do not convey any statistical significance and can be subjectively interpreted. Interpretation of these plots should always be subject to subsequent statistical analysis.

```{r community-composition-plots - Cecal content only}
ps1.cecal.rarefied.mpse
```

```{r}
# Calculate relative abundance
ps1.cecal.rarefied.mpse %<>%
    mp_cal_abundance( # for each samples
      .abundance = Abundance
    ) %>%
    mp_cal_abundance( # for each groups 
      .abundance=Abundance,
      .group=Treatment
    )
ps1.cecal.rarefied.mpse
```



```{r}
# visualize the relative abundance of top 20 phyla for each .group (time)
colors <- c('#F7911E', '#767BBB', '#908E35', '#915323', '#FDD379', '#D1D1CC','#3D57A7') #, '#EE3324', '#F7921E',  '#F6EC21', '#79797A')
p3 <- ps1.cecal.rarefied.mpse %>%
         mp_plot_abundance(
            .abundance=RelAbundance, 
            .group=Treatment,
            taxa.class = Phylum,
            topn = 100,
            plot.group = TRUE,
          ) +
    scale_fill_manual(values = colors) +
    ggtitle("Phyla Relative Abundance of Samples From Cecal Content") +
    theme(legend.position = "right", legend.text = element_text(size = 12), axis.text.x = element_text(size = 10, angle = 0, hjust = 0.5))
pdf("./Plots/Phylum_relab_Cecal.plots.pdf")
p3
dev.off()
```

```{r}
ps1.cecal.rarefied
# Extract relevant data from phyloseq object
taxa_table <- phyloseq::tax_table(ps1.cecal.rarefied)
sample_data <- sample_data(ps1.cecal.rarefied)

```

```{r}
# Create a data table for ggplot
phylum_ps1.cecal.rarefied <- ps1.cecal.rarefied %>% tax_glom(taxrank = "Phylum") %>% 
        transform_sample_counts(function(x) {x/sum(x)}) %>% psmelt()
phylum_ps1.cecal.rarefied

# Filter data for Bacteroidetes , Firmicutes , Proteobacteria, Verrucomicrobia
bacteroidetes_data <- phylum_ps1.cecal.rarefied %>%
  filter(Phylum == "Bacteroidetes")

b_data_ps <- subset_taxa(ps1.cecal.rarefied, Phylum=="Bacteroidetes")

firmicutes_data <- phylum_ps1.cecal.rarefied %>%
  filter(Phylum == "Firmicutes")

proteobacteria_data <- phylum_ps1.cecal.rarefied %>%
  filter(Phylum == "Proteobacteria")

verrucomicrobia_data <- phylum_ps1.cecal.rarefied %>%
  filter(Phylum == "Verrucomicrobia")


# Sum abundance for each sample for each phylum
bacteroidetes_sum <- summarise(group_by(bacteroidetes_data, Sample, Treatment), Abundance = sum(Abundance))
firmicutes_sum <- summarise(group_by(firmicutes_data, Sample, Treatment), Abundance = sum(Abundance))
proteobacteria_sum <- summarise(group_by(proteobacteria_data, Sample, Treatment), Abundance = sum(Abundance))
verrucomicrobia_sum <- summarise(group_by(verrucomicrobia_data, Sample, Treatment), Abundance = sum(Abundance))


# Merge the two data frames based on the Sample column
merged_data_bf <- merge(bacteroidetes_sum, firmicutes_sum, by = "Sample", suffixes = c("_Bacteroidetes", "_Firmicutes"))
merged_data_pf <- merge(proteobacteria_sum, firmicutes_sum, by = "Sample", suffixes = c("_Proteobacteria", "_Firmicutes"))
merged_data_vf <- merge(verrucomicrobia_sum, firmicutes_sum, by = "Sample", suffixes = c("_Verrucomicrobia", "_Firmicutes"))


# Calculate the Bacteroidetes/Firmicutes ratio
merged_data_bf$Bact_Firm_Ratio <- merged_data_bf$Abundance_Bacteroidetes / merged_data_bf$Abundance_Firmicutes
# Calculate the Proteobacteria/Firmicutes ratio
merged_data_pf$Prot_Firm_Ratio <- merged_data_pf$Abundance_Proteobacteria / merged_data_bf$Abundance_Firmicutes
# Calculate the Verrucomicrobia/Firmicutes ratio
merged_data_vf$Verr_Firm_Ratio <- merged_data_vf$Abundance_Verrucomicrobia / merged_data_bf$Abundance_Firmicutes

```


```{r}
# Perform Wilcoxon rank-sum test
wilcox_test_result_bf <- wilcox.test(Bact_Firm_Ratio ~ Treatment_Firmicutes, data = merged_data_bf)

# Calculate the midpoint between boxplots
midpoint <- 0.5 + 1:length(unique(merged_data_bf$Treatment_Firmicutes))

# Create the ggplot
p_bf <- ggplot(merged_data_bf, aes(x = Treatment_Firmicutes, y = Bact_Firm_Ratio, fill = Treatment_Firmicutes)) +
  geom_boxplot() +
  labs(x = NULL, y = "Bacteroidetes/Firmicutes") +
  theme_minimal() + 
  scale_fill_manual(name = "Treatment", values = c("steelblue", "firebrick")) +
  
  # Add p-value annotation
  annotate("text", x = 1.5, y = 1.5,
           label = sprintf("p = %.4f", wilcox_test_result_bf$p.value),
           hjust = 0.5, vjust = 0, size = 5)
pdf('./Plots/Cecal_Bact_Firm_Ratio_phy.pdf')
p_bf
dev.off()

```



```{r}

# Create a boxplot Proteobacteria/Firmicutes 
wilcox_test_result_pf <- wilcox.test(merged_data_pf$Prot_Firm_Ratio ~ Treatment_Firmicutes, data = merged_data_pf)
p_pf <- ggplot(merged_data_pf, aes(x = Treatment_Firmicutes, y = Prot_Firm_Ratio, fill = Treatment_Firmicutes)) +
  geom_boxplot() +
  labs(x=NULL, y = "Proteobacteria/Firmicutes") +
  theme_minimal() + 
  scale_fill_manual(name = "Treatment", values = c("steelblue", "firebrick")) +
  
  # Add p-value annotation
  annotate("text", x = 1.5, y = 0.06,
           label = sprintf("p = %.4f", wilcox_test_result_pf$p.value),
           hjust = 0.5, vjust = 0, size = 5)
pdf('./Plots/Cecal_Prot_Firm_Ratio_phy.pdf')
p_pf
dev.off()
```


```{r}
# Create a boxplot Verr_Firm
wilcox_test_result_vf <- wilcox.test(merged_data_vf$Verr_Firm_Ratio ~ Treatment_Firmicutes, data = merged_data_vf)
p_vf <- ggplot(merged_data_vf, aes(x = Treatment_Firmicutes, y = Verr_Firm_Ratio, fill = Treatment_Firmicutes)) +
  geom_boxplot() +
  labs(x=NULL, y = "Verrucomicrobia/Firmicutes") +
  theme_minimal() + 
  scale_fill_manual(name = "Treatment", values = c("steelblue", "firebrick")) +
  
  # Add p-value annotation
  annotate("text", x = 1.5, y = 1,
           label = sprintf("p = %.4f", wilcox_test_result_vf$p.value),
           hjust = 0.5, vjust = 0, size = 5)
pdf('./Plots/Cecal_Verr_Firm_Ratio_phy.pdf')
p_vf
dev.off()
```





#### 3.6 Biomarker discovery


```{r}
ps1.cecal.rarefied.mpse %>% print(width=150)
```


```{r}
genus.tb <- ps1.cecal.rarefied.mpse %>% 
            mp_extract_abundance(taxa.class=Genus)
# The abundance of each samples is nested, it can be flatted using the unnest of tidyr.
genus.tb %<>% tidyr::unnest(cols=AbundanceBySample) %>% dplyr::rename(Genus="label")
genus.tb

ps1.cecal.rarefied.mpse %<>%
    mp_diff_analysis(
       .abundance = AbundanceBySample,
       .group = Treatment,
       first.test.alpha = 0.01)
# The result is stored to the taxatree or otutree slot, you can use mp_extract_tree to extract the specific slot.
taxa.tree <- ps1.cecal.rarefied.mpse %>% 
               mp_extract_tree(type="taxatree")
taxa.tree
```







### Betadisper and permutational ANOVA

#### As we saw earlier, we have some information about our samples in our sample info table. There are many ways to incorporate this information, but one of the first I typically go to is a permutational ANOVA test to see if any of the available information is indicative of community structure. Here we are going to test if there is a statistically signficant difference between our sample types. One way to do this is with the betadisper and adonis functions from the vegan package. adonis can tell us if there is a statistical difference between groups, but it has an assumption that must be met that we first need to check with betadisper, and that is that there is a sufficient level of homogeneity of dispersion within groups. If there is not, then adonis can be unreliable.

## Group significance testing with ADONIS 

```{r adonis-script}
# Set a random seed so that exact results can be reproduced
set.seed(10000)

# Function to run adonis test on a physeq object and a variable from metadata 
doadonis <- function(physeq, category) {
  bdist <- phyloseq::distance(physeq, "unifrac")
  col <- as(sample_data(physeq), "data.frame")[ ,category]
  
  # Adonis test
  adonis.bdist <- adonis(bdist ~ col)
  print("Adonis results:")
  print(adonis.bdist)
  
}

```

```{r}

permanova_cecal <- doadonis(ps1.cecal.rarefied, "Treatment")
write.table(permanova_cecal$aov.tab, "./permanova_cecal.txt")
```


#### That said, recovered 16S rRNA gene copy numbers do represent… well, numbers of recovered 16S rRNA gene copies. So long as you’re interpreting them that way, and thinking of your system in the appropriate way, you can perform differential abundance testing to test for which representative sequences have significantly different copy-number counts between samples – which can be useful information and guide the generation of hypotheses. One tool that can be used for this is DESeq2, which we used above to transform our count table for beta diversity plots.

##### Use microbioMarker package to run aldex2. 



```{r}
# Genus Level
genus_ps1.cecal.rarefied <- ps1.cecal.rarefied %>% tax_glom(taxrank = "Genus") 

mm_aldex_genus <- run_aldex(
     genus_ps1.cecal.rarefied,
     group = 'Treatment',
     method = "glm_anova", # Generalized linear model (GLM) approach for comparing means. Appropriate for non-normally distributed count data, especially when dealing with microbiome data.
     mc_samples = 128 )

```

```{r}
markers <- marker_table(mm_aldex_genus)$feature # extract the tip labels from feature table

pdf('./Plots/Cecal_aldex_2_Genus_cladogram.pdf')
cladogram <- plot_cladogram(mm_aldex_genus, only_marker = TRUE, color = c("steelblue", "firebrick"), alpha=0.5, branch_size = 0.5, clade_label_font_size = 7) +
theme(plot.margin = margin(0, 0, 0, 0))  

cladogram + geom_tiplab(label=markers, size=1.5, aes(angle=angle))
dev.off()


```


```{r}
pdf('./Plots/Cecal_aldex_heatmap_genus.pdf')
plot_heatmap(mm_aldex_genus, transform = "log10p", group = "Treatment",   cluster_marker = TRUE, max_label_len = 60, sample_label = TRUE, annotation_col = NULL)
dev.off()

```

### Plot relative abundance for cecal content
```{r}
# Create a data table for ggplot

colors <- c('#F0C18A','#008D47','blue','#00AEAA', '#767BBB', '#908E35', '#915323', '#FDD379', '#D1D1CC','#3D57A7', '#00ABD2', 
            '#F5EB00', '#A0489B','#F26822','#F175AD', '#919799', '#B52026', '#BCAED5', '#3E8386','#CCCC64' ,
            '#5793CE', '#8D3E1E', '#6591CC', '#986627', '#4D622C', '#D75345', '#483987' , '#951B53', '#696A6C')



# Create a data table for ggplot
ps1_genus.0.005 <- ps1.cecal.rarefied %>%
  tax_glom(taxrank = "Genus") %>%                      # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance (or use ps0.ra)
  psmelt() %>%                                         # Melt to long format for easy ggploting
  filter(Abundance > 0.005)                             # Filter out low abundance taxa

# Plot - Genus

p.ps1_genus.0.005 <- ggplot(ps1_genus.0.005, aes(x = Sample_name, y = Abundance, fill = Genus)) + 
  geom_bar(stat = "identity", width = 0.8, ) +
  facet_wrap(~Treatment, scales = "free_x", nrow = 4, ncol = 7) +
  theme(axis.text.x = element_blank()) +
  theme(axis.title.x = element_blank()) +
  labs(title = "Realtive Abundant Genus in Cecal Content (> 0.5%)") + 
  scale_fill_manual(values=colors) +
  theme(
    legend.key.size = unit(0.3, "cm"),   # Reduce the size of the legend key (box)
    legend.key.height = unit(0.3, "cm"), 
    legend.text = element_text(size = 10),   # Adjust the size of the legend text
    axis.line = element_line(size = 0.3),   # Set size of x and y axis lines
    panel.grid.major = element_blank(),     # Remove major grid lines
    panel.grid.minor = element_blank()    # Remove minor grid lines
  ) + theme(aspect.ratio=1/0.5)
p.ps1_genus.0.005


# You can rerun the first bit of code in this chunk and change Phylum to Species, Genus, etc.


pdf("./Plots/Cecal_relative_plot_genus_0.5.pdf")
print(p.ps1_genus.0.005)
dev.off()


colors <- c('#F0C18A','#008D47','#767BBB', '#00AEAA','blue', '#908E35', '#915323', '#FDD379', '#D1D1CC','#3D57A7', '#00ABD2', 
            '#F5EB00', '#A0489B','#F26822','#F175AD', '#919799', '#B52026', '#BCAED5', '#3E8386','#CCCC64' ,
            '#5793CE', '#8D3E1E', '#6591CC', '#986627', '#4D622C', '#D75345', '#483987' , '#951B53', '#696A6C')


# Create a data table for ggplot
ps1_genus.0.001 <- ps1.cecal.rarefied %>%
  tax_glom(taxrank = "Genus") %>%                      # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance (or use ps0.ra)
  psmelt() %>%                                         # Melt to long format for easy ggploting
  filter(Abundance > 0.001)                             # Filter out low abundance taxa

write.table(ps1_genus.0.001, "./Tables/Cecal_content_relativeAbundance_genus.tsv", sep='\t', quote=F, col.names = F, row.names = F)

# Plot - Genus

p.ps1_genus.0.001 <- ggplot(ps1_genus.0.001, aes(x = Sample_name, y = Abundance, fill = Genus)) + 
  geom_bar(stat = "identity", width = 0.8, ) +
  facet_wrap(~Treatment, scales = "free_x", nrow = 4, ncol = 7) +
  theme(axis.text.x = element_blank()) +
  theme(axis.title.x = element_blank()) +
  labs(title = "Realtive Abundant Genus in Cecal Content (> 0.1%)") + 
  scale_fill_manual(values=colors) +
  theme(
    legend.key.size = unit(0.3, "cm"),   # Reduce the size of the legend key (box)
    legend.key.height = unit(0.3, "cm"), 
    legend.text = element_text(size = 10),   # Adjust the size of the legend text
    axis.line = element_line(size = 0.3),   # Set size of x and y axis lines
    panel.grid.major = element_blank(),     # Remove major grid lines
    panel.grid.minor = element_blank()    # Remove minor grid lines
  ) + theme(aspect.ratio=1/0.5)
p.ps1_genus.0.001


# You can rerun the first bit of code in this chunk and change Phylum to Species, Genus, etc.


pdf("./Plots/Cecal_relative_plot_genus_0.1.pdf")
print(p.ps1_genus.0.001)
dev.off()

```

